### PR TITLE
fix: detect companion host-port collisions at dispatch time (HTTP 409) (#304)

### LINKED ISSUE #304: Companion host-port mappings collide silently across concurrent workspaces — detect at dispatch time
## Problem

A companion service can declare `ports: [[container_port, host_port]]` in its dispatch payload (e.g., `[[8000, 18000]]`). When TWO concurrent workspaces both have a companion mapping the SAME `host_port`, the second one's compose-up fails at container start:

```
Error response from daemon: failed to set up container networking: driver failed programming external connectivity
on endpoint awf-ws_<id>-aira-agent: Bind for 0.0.0.0:18000 failed: port is already allocated
```

AWF doesn't detect the collision at dispatch time. The second workspace flips to `INFRASTRUCTURE_FAILURE` after build + stack-up, wasting:
- The companion image build time (~3-5 min for aira-agent)
- The full DinD + postgres provisioning
- One full agent-runtime slot
- LLM credits if the agent was about to start

The operator has no way to know in advance that a port is taken — the AWF API accepts the dispatch payload happily and the failure only surfaces 4-5 min in.

## Reproduction

1. Dispatch workspace A with `companions: [{ports: [[8000, 18000]], ...}]` — succeeds.
2. While A is still alive (running, monitoring_pr, comment_repair — any non-terminal state), dispatch workspace B with the same `ports: [[8000, 18000]]`.
3. B passes profile validation, claims a worker, builds the companion image, attempts compose-up, fails at container networking, terminates with `INFRASTRUCTURE_FAILURE` after ~5 min.

In my case: T170 dispatch (`ws_75cb19127ecc43b1909d3e24`) collided with T160 (`ws_84a0d4894cb4404997172398`) on host port 18000 since both reused the same companion env from a prior snapshot.

## Workaround

Drop the `ports` array entirely from the companion config. The companion is reachable from inside the workspace network via Docker service-name DNS (`http://aira-agent:8000`) — host port mapping is only useful for the OPERATOR to curl from their laptop, not for the agent's tests. Removing it eliminates collisions across all parallel workspaces.

But this is a workaround, not a contract. Operators who genuinely want host-port access (e.g., to attach a debugger) shouldn't have to silently coordinate ports across concurrent dispatches.

## Proposed shape

Detect the collision at one of three places (ranked by intervention cost):

**Cheap — dispatch-time validation:**
- When `POST /v1/workspaces` accepts a payload with `companions[].ports`, check the AWF state DB for any non-terminal workspace whose companion declares the same `host_port` (same operator/node — port namespace is per-host).
- If found, return HTTP 409 with a clear message: `host port {N} is already mapped by workspace {other_id}`.
- This is a single SELECT against the workspace table — negligible cost, catches 100% of operator-side collisions, NO build/provision waste.

**Medium — auto-allocate ephemeral ports:**
- If `host_port == null` (or a new sentinel like `auto`), AWF allocates a free host port at dispatch and stores it in the workspace's state for the operator to look up via `GET /v1/workspaces/{id}/runtime`.
- The companion still gets mapped, but the operator stops thinking about ports.

**Premium — also surface the collision in the failure event payload:**
- When `INFRASTRUCTURE_FAILURE` fires for a port-bind error, parse the docker daemon's error string, extract the port + colliding workspace ID, and embed them in the event payload. Today the message is buried in worker logs (operator has to `docker logs awf-local-service-worker-1` to find it).
- Pairs well with #299 (companion logs in failure event).

My preference: ship the dispatch-time check (cheap), and the auto-allocation flag (`host_port: null` → ephemeral) as a follow-up.

## Related

- #289 (companion services contract) — this is a usability gap, not a contract bug.
- #299 (opaque failure events) — same class of "failure mode is buried in worker logs not surfaced to operator".

### PR DESCRIPTION
**Pull request by dimileeh**
## What

Dispatch-time detection of companion host-port collisions (issue #304, "Cheap" option): `POST /v1/workspaces` now returns **HTTP 409 `HOST_PORT_CONFLICT`** when a companion's `host_port` is already mapped by a non-terminal workspace — before any build/provision, instead of a ~5-min `INFRASTRUCTURE_FAILURE` at compose-up.

## Changes

- `db/repositories/base.py` — `HostPortConflict` dataclass.
- `db/repositories/workspace_repo.py` — `find_host_port_conflicts(host_ports, excluding_workspace_id=None)`: scans only `ACTIVE_OWNED_PATH_OVERLAP_STATUSES` (non-terminal) workspaces, parses `task_policy.companions[].ports`, returns conflicts. TOCTOU documented.
- `service/workspaces.py` — `WorkspaceCreateHostPortConflictError` (`error_code="HOST_PORT_CONFLICT"`); `WorkspaceService.create()` checks for conflicts after the disk check.
- `api/routes/workspaces.py` — maps the error to 409 `ErrorResponse`.
- 11 unit tests: collision, no-collision, terminal-not-blocking (all 5 terminal statuses), multi-port, multi-companion, idempotent replay (`excluding_workspace_id`), empty/no-companions/no-ports edge cases.

## Out of scope (deferred follow-ups, per the issue)

- Auto-allocating ephemeral host ports (`host_port: null` → `auto`).
- Surfacing the collision in `INFRASTRUCTURE_FAILURE` event payloads (overlaps #299).

## Design notes for reviewers

- Conflict detection is **host-wide** (not per-repo): host ports are a per-host namespace, so a collision is independent of repo/branch.
- JSON is parsed in **Python** rather than dialect-specific JSONB/`json_extract` SQL — works identically on Postgres and SQLite, small query volume (only non-terminal rows). Open to switching to SQL if preferred.

## Validation (re-run independently before opening)

`ruff check .` ✅ · `ruff format --check .` ✅ · `mypy` (288 files) ✅ · `generate_openapi.py --check` ✅ · targeted suite 11/11 ✅.

## Provenance

This is the work product of AWF dogfood workspace `ws_b815524e9ae5413ba6216729` (opencode/glm-5.1:cloud, xhigh). That workspace was marked `failed` not because the implementation was wrong, but because the agent committed its own changes during a pre-commit repair pass, which broke AWF's post-agent commit handshake — see #327. The branch was salvaged and pushed as-is; the implementation is complete and validated.

Closes #304.

---

> [!NOTE]
> **High Risk**
> Changes span workspace admission, retry gates, provisioner launch semantics, and cleanup/planning auto-retry concurrency—areas where incorrect port or runtime-release logic can admit colliding Compose stacks or strand retries.
> 
> **Overview**
> This PR extends **dispatch-time host-port admission** beyond companions to **profile service ports**, with **HTTP 409** responses on create and retry (`HOST_PORT_CONFLICT`, `DUPLICATE_HOST_PORT`, `SOURCE_RUNTIME_NOT_RELEASED`) and matching **reason catalog** / OpenAPI descriptions.
> 
> **Conflict detection** now considers active and terminal-but-unreleased workspaces: companions, **resolved** profile ports, and **`requested_profile` fallback** when resolution has not run; it respects **node scope**, **terminal-runtime release** events (with deterministic **`WorkspaceEvent.id`** tie-breaking), and legacy rows (null compose metadata, reservations, node-stamped failures). **Advisory locking** runs before retry source-runtime checks to close admission races.
> 
> **Retry** is blocked until the source runtime is effectively released or pre-launch failure is proven; **provisioner** only marks compose as launched via an **`on_compose_up_started`** callback so pre-compose failures do not pin ports. **Cleanup workers** use **effective local node id**, record **`workspace.terminal_runtime_released`**, and drive a **planning-scope auto-retry** state machine (blocked on port/source-runtime conflicts, deduped events, resume after release, manual-retry guards). Worker **claims** add node-scope conditions; large test modules were split for the line limit; **development** merge conflicts were resolved in catalog and PR adoption code.
> 
> A **pre-GKE** backlog item notes indexed host-port admission for scale.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6a53656a299a98e81238e4621ef7c5b4be667ba7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added host-port conflict detection and validation during workspace creation and retry to prevent port collisions.
  * Introduced terminal runtime release event tracking for workspaces to improve lifecycle management.

* **Bug Fixes**
  * Improved host-port admission handling with earlier conflict detection and consistent error responses.
  * Enhanced workspace retry logic with better source workspace runtime release validation.

* **Documentation**
  * Added reason catalog entries for host-port conflicts, duplicate port claims, and source runtime release issues.

### CODE REVIEW

**Comment by dimileeh** [comment]
AWF deferred the review thread on `src/awf/api/routes/workspaces.py` and filed https://github.com/dimileeh/aira-agent-workspace-fabric/issues/347 to track the follow-up. Resolving this thread; the deferred work lives in that issue.

**Comment by dimileeh** [comment]
Operator note on the latest Greptile review (3 issues), resolved at `3ecb6980`:

- **Issue 1 — provisioner static-profile bypass** (`node/provisioner.py`): already fixed earlier on this branch. `compose_project_name` is now persisted under a row lock (`get_for_update`) in the pre-launch block *before* `launch`, outside the `if profile_resolution is not None:` guard, so it runs for static profiles too. Anchor commit was stale.
- **Issue 2 — cancel release-event** (`service/controls.py`): not a present bug. The `terminal_runtime_released` event is only recorded after a successful `compose stop`; a failed stop re-raises `WorkspaceStackStopError` before any release event, and the event is guarded by `compose_project_name is not None` + terminal-status + no-prior-release. Anchor commit was stale.
- **Issue 3 — `compose_project_name IS NOT NULL` guard** (`db/repositories/workspace_repo_host_ports.py`): correct by design, now documented + pinned. The provisioner persists `compose_project_name` under a row lock before launching the stack and never persists it for pre-launch failures, so among terminal workspaces with no `terminal_runtime_released` event, `NULL ⟺ no containers ⟺ no ports held`. Widening the guard to also match `NULL` rows (the literal suggestion) would falsely flag never-launched workspaces as port holders and regress reuse. This commit adds an inline invariant comment cross-referencing the provisioner, plus a combined regression test (`test_compose_project_name_null_invariant_distinguishes_port_holders`) asserting both halves so any such widening fails loudly.

No behavior change in this commit — documentation + regression guard only.

File: src/awf/db/repositories/workspace_repo_host_ports.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        stmt = stmt.where(
+            or_(
+                effective_node == node_id,
+                and_(
+                    Workspace.node_id.is_(None),
+                    active_reservation_node.is_(None),
+                    latest_reservation_node.is_(None),
+                ),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories/workspace_repo_host_ports.py:356]: @cursor False positive — Workspace.node_id is never a container hostname.

It's resolved to a stable configured value: effective_worker_node_id / effective_service_node_id (src/awf/service/node_identity.py) return settings.worker_node_id / settings.node_id if set, else the constant DEFAULT_LOCAL_SERVICE_WORKER_NODE_ID (e.g. local). Every stamp site uses self._config.node_id / effective_worker_config_node_id(self._config), never socket.gethostname(). So all rows on one machine carry the same stable id, which matches effective_node == node_id exactly — nothing is omitted.

The exact-node match is also deliberate and required for multi-node: host ports are node-local, so a port holder on node B must not block a create on node A. Broadening the filter to also include non-matching-node rows (so nothing can be "missed") would manufacture false-positive HOST_PORT_CONFLICTs in a real multi-node deployment — the opposite of correct.

The only residual edge — a row carrying a stale node_id after worker_node_id is reconfigured — is already documented in this function's NODE-FILTER FALLBACK docstring section as future work (a claimed_node_id column for precise attribution), and the null-placement fallbacks already cover the legacy no-node/no-reservation case. No code change.

File: src/awf/control/executor/planning_ops.py [hunk 2]
```diff
@@ -528,52 +574,328 @@ async def _auto_retry_planning_scope_failure(
...
+        )
+        .limit(_PLANNING_SCOPE_AUTO_RETRY_TERMINAL_RELEASE_SCAN_LIMIT)
+    )
+    events = list((await session.execute(stmt)).scalars())
+    for event in events:
+        event_type = getattr(event, "event_type", None)
+        payload = _planning_scope_auto_retry_payload(event)
+        if _planning_scope_auto_retry_terminal_release_event_matches(event_type, payload):
+            return event
+    return None
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/executor/planning_ops.py:876]: Deferred to the `event_order` backfill follow-up — this is the same narrow legacy edge as the cleanup query's tiebreaker (and the earlier `newer_planning_event_exists` discussion).

`event_order` is assigned sequentially on every new event (`WorkspaceRepository._reserve_workspace_event_orders`), so `id.desc()` only acts as a tiebreaker when two events share an exact-microsecond `occurred_at` **and** both have `NULL event_order` — i.e. only rows created before the `event_order` column existed. Modern rows are fully deterministic via `occurred_at.desc(), event_order.desc().nullslast()`.

Removing `id.desc()` doesn't fix the legacy case: it replaces a deterministic-but-time-meaningless order with a **non-deterministic** one (no tiebreaker → Postgres-arbitrary), which is worse for reproducibility and still picks an arbitrary event in the tie. The correct fix is the backfill migration that populates `event_order` for legacy rows (and makes the column `NOT NULL`), making this branch dead. Tracked as a follow-up; keeping `id.desc()` preserves deterministic ordering meanwhile.

File: src/awf/control/worker/cleanup.py [hunk 3]
```diff
@@ -162,6 +187,19 @@ async def _release_terminal_runtime_resources(self: Any) -> None:
...
+        raise
+    except Exception as exc:
+        if not release_errors:
+            raise
+        _log.warning(
+            "worker.terminal_runtime_release_resume_scan_failed_after_release_error",
+            reason_code=_PLANNING_SCOPE_AUTO_RETRY_RESUME_FAILED_REASON_CODE,
+            error_type=type(exc).__name__,
+            error=str(exc)[:240],
+        )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/control/worker/cleanup.py:202]: Not an issue — the resume is idempotent by design, so a second pass is a harmless recheck, not a double-retry/double-creation.

`_resume_blocked_planning_scope_auto_retry_after_runtime_release` records a **deduplicated** host-port block and leaves the workspace a candidate only while it's still blocked (see the `_resume_pending_planning_scope_auto_retries_after_terminal_release` docstring: *"records a deduplicated host-port block and remains a candidate. The next cleanup scans intentionally keep rechecking until the port is free"*). So:

- If the targeted resume (654-668) **succeeds**, it clears the pending marker; that commit lands before the batch sweep's candidate query (190-202) in the same sequential cleanup task, so the sweep no longer lists the workspace — no second `retry_workspace_row()`, no double-creation.
- If it's **still blocked**, the marker stays and the sweep rechecks it — but that's the intended recheck loop, and the dedup guard makes it a no-op on state.

Net: at most one extra idempotent recheck per cycle for a still-blocked workspace, never a duplicate retry. No fix needed.

File: src/awf/node/provisioner_helpers.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        "omitted_optional_count",
+        "omitted_optional",
+        "skipped_unresolved_count",
+        "companion_env_secret_count",
+        "companion_env_secrets",
+        "companion_omitted_optional_env_secret_count",
+        "companion_omitted_optional_env_secrets",
+    ):
+        if key in plan_metadata:
+            metadata[key] = plan_metadata[key]
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/node/provisioner_helpers.py:61]: Valid — fixed in follow-up PR #359.

`_stack_secret_lease_mount_metadata` now redacts `companion_*` fields via `redact_audit_value`, mirroring `_stack_companion_env_secret_event_payload`, so both event paths are consistent with AWF's redaction-mandatory rule. Split into a separate PR (rather than blocking this already-green #304 head another full CI cycle) since the fields are metadata-only today (defense-in-depth, no active leak). #359 includes a regression test asserting the mount-metadata event redacts a sensitive companion field identically to the companion-secret event.
